src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs

```diff
@@ -187,30 +184,22 @@ private void RestorePackages()
                 });
         }
 
-        public override async Task Invoke(object[] parameters)
+        public override async Task InvokeInternal(object[] parameters)
         {
-            FunctionStartedEvent startedEvent = null;
-            string invocationId = null;
-
-            try
-            {
-                ExecutionContext functionExecutionContext = (ExecutionContext)parameters.First(p => p is ExecutionContext);
-                invocationId = functionExecutionContext.InvocationId.ToString();
-
-                // Separate any system parameters we might have added from the actual method parameters
-                object[] originalParameters = parameters;
-                MethodInfo function = await GetFunctionTargetAsync();
-                int actualParameterCount = function.GetParameters().Length;
-                object[] systemParameters = parameters.Skip(actualParameterCount).ToArray();
-                parameters = parameters.Take(actualParameterCount).ToArray();
-
-                startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
-                _metrics.BeginEvent(startedEvent);
 
-                TraceWriter.Info(string.Format("Function started (Id={0})", invocationId));
+            // Separate system parameters from the actual method parameters
+            object[] originalParameters = parameters;
+            MethodInfo function = await GetFunctionTargetAsync();
+            int actualParameterCount = function.GetParameters().Length;
+            object[] systemParameters = parameters.Skip(actualParameterCount).ToArray();
+            parameters = parameters.Take(actualParameterCount).ToArray();
 
-                parameters = ProcessInputParameters(parameters);
+            parameters = ProcessInputParameters(parameters);
 
+            // We don't know if this function is sync or async so wrap it in a Task. 
+            // Otherwise, a sync function with an infinite loop prevents us from ever cancelling.
```

Comment:
Rather than Task.Run, can't we just start a timer before invoking the user code, and stop it when the function is done? If the timer expires, it triggers cancellation, and that would take effect regardless of whether the user code blocks.